### PR TITLE
[Fix] Allow mixed precision training with SimOTAAssigner

### LINKED ISSUE #7474: SimOTAAssigner does not support half precision training
Thanks for your error report and we appreciate it a lot.

**Checklist**

1. I have searched related issues but cannot get the expected help.
2. I have read the [FAQ documentation](https://mmdetection.readthedocs.io/en/latest/faq.html) but cannot get the expected help.
3. The bug has not been fixed in the latest version.

**Describe the bug**

Trying to run YOLOX with the SimOTAAssigner in 16-bit precision mode, but it fails with a `RuntimeError: "sqrt" "_vml_cpu" not implemented for 'Half'`.

**Reproduction**

1. What command or script did you run?

```none
A placeholder for the command.
```

2. Did you make any modifications on the code or config? Did you understand what you have modified?
3. What dataset did you use?

**Environment**

1. Please run `python mmdet/utils/collect_env.py` to collect necessary environment information and paste it here.
```
No CUDA runtime is found, using CUDA_HOME='/usr/local/cuda'
sys.platform: linux
Python: 3.9.9 | packaged by conda-forge | (main, Dec 20 2021, 02:41:03) [GCC 9.4.0]
CUDA available: False
GCC: gcc (GCC) 9.4.0
PyTorch: 1.10.0
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2022.0-Product Build 20211112 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

TorchVision: 0.11.1+cu102
OpenCV: 4.5.3
MMCV: 1.4.6
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 11.3
MMDetection: 2.22.0+d939bb6
```
2. You may add addition that may be helpful for locating the problem, such as
    - How you installed PyTorch [e.g., pip, conda, source] **conda**
    - Other environment variables that may be related (such as `$PATH`, `$LD_LIBRARY_PATH`, `$PYTHONPATH`, etc.)

**Error traceback**
If applicable, paste the error trackback here.

```python-traceback
File "/home/username/Documents/github/project/project/yoloxnet.py", line 240, in evaluate
  loss_dict: dict = self.yolox_head.loss(
File "/home/username/miniconda/envs/project/lib/python3.9/site-packages/mmcv/runner/fp16_utils.py", line 197, in new_func
  return old_func(*args, **kwargs)
File "/home/username/miniconda/envs/project/lib/python3.9/site-packages/mmdet/models/dense_heads/yolox_head.py", line 381, in loss
  num_fg_imgs) = multi_apply(
File "/home/username/miniconda/envs/project/lib/python3.9/site-packages/mmdet/core/utils/misc.py", line 30, in multi_apply
  return tuple(map(list, zip(*map_results)))
File "/home/username/miniconda/envs/project/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
  return func(*args, **kwargs)
File "/home/username/miniconda/envs/project/lib/python3.9/site-packages/mmdet/models/dense_heads/yolox_head.py", line 461, in _get_target_single
  assign_result = self.assigner.assign(
File "/home/username/miniconda/envs/project/lib/python3.9/site-packages/mmdet/core/bbox/assigners/sim_ota_assigner.py", line 85, in assign
  assign_result = self._assign(pred_scores, priors, decoded_bboxes,
File "/home/username/miniconda/envs/project/lib/python3.9/site-packages/mmdet/core/bbox/assigners/sim_ota_assigner.py", line 161, in _assign
  valid_pred_scores.sqrt_(), gt_onehot_label,
RuntimeError: "sqrt" "_vml_cpu" not implemented for 'Half'
```

**Bug fix**
If you have already identified the reason, you can provide the information here. If you are willing to create a PR to fix it, please also leave a comment here and that would be much appreciated!

Requires modifying the code somewhere here:

https://github.com/open-mmlab/mmdetection/blob/52a32763bef402a5eb684625d974eebe29755351/mmdet/core/bbox/assigners/sim_ota_assigner.py#L160-L162

The `.sqrt_()` doesn't work with `torch.float16` on CPU, and also, the binary_cross_entropy function seems to require float32 inputs.

Related to https://github.com/pytorch/pytorch/issues/50789

### PR DESCRIPTION
**Pull request by weiji14**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Enable 16-bit precision training on YOLOX with the SimOTAAssigner. Fixes #7474.

## Modification

Cast the `valid_pred_scores` variable to float32 dtype in the `F.binary_cross_entropy` calculation.

Also added a unit test for SimOTAAssigner, though it's not exactly a good one. Wasn't sure how best to test this properly.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by RangiLyu — APPROVED** [review]

File: mmdet/core/bbox/assigners/sim_ota_assigner.py [hunk 1]
```diff
@@ -158,7 +158,7 @@ def _assign(self,
 
         valid_pred_scores = valid_pred_scores.unsqueeze(1).repeat(1, num_gt, 1)
         cls_cost = F.binary_cross_entropy(
-            valid_pred_scores.sqrt_(), gt_onehot_label,
+            valid_pred_scores.to(dtype=torch.float32).sqrt_(), gt_onehot_label,
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/sim_ota_assigner.py:161]: should we convert cls_cost back to the same type of valid_pred_scores after calculating it to float32?

Comment (weiji14) [reply mmdet/core/bbox/assigners/sim_ota_assigner.py:161]: You mean something like this?

```suggestion
            valid_pred_scores.to(dtype=torch.float32).sqrt_(), gt_onehot_label,
        cls_cost.to(dtype=valid_pred_scores.dtype)
```

Comment (ZwwWayne) [reply mmdet/core/bbox/assigners/sim_ota_assigner.py:161]: yes. Is it necessary?

**Comment by ZwwWayne** [comment]
Please also fix lint issues so that we can merge this PR.

File: mmdet/core/bbox/assigners/sim_ota_assigner.py (hunk 1 above)
Comment (weiji14) [reply mmdet/core/bbox/assigners/sim_ota_assigner.py]: Seems to run fine with or without casting back to float16. I guess we could just convert it back then, done in c3f917172c04560baed9373d0fdc9338411bc94d.

File: mmdet/core/bbox/assigners/sim_ota_assigner.py [hunk 2]
```diff
@@ -158,8 +158,11 @@ def _assign(self,
 
         valid_pred_scores = valid_pred_scores.unsqueeze(1).repeat(1, num_gt, 1)
         cls_cost = F.binary_cross_entropy(
-            valid_pred_scores.sqrt_(), gt_onehot_label,
-            reduction='none').sum(-1)
+            valid_pred_scores.to(dtype=torch.float32).sqrt_(),
+            gt_onehot_label,
+            reduction='none',
+        ).sum(-1)
+        cls_cost.to(dtype=valid_pred_scores.dtype)
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/sim_ota_assigner.py:165]: It is better to write
```
.sum(-1).to(dtype=valid_pred_scores.dtype)
```

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**